### ISSUE TITLE
Make rebuild.latestClass more compliant with new style classes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2297]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2297|
|Type|defect|
|Created|2006-12-14 15:04:13Z|

test_rebuild has a todo for new styles classes for a long time.

I attach a patch correcting the todo; nothing too smart here, I just skip latestClass for !__builtin!__ classes, which looked reasonable to me.

Attachments:

* [rebuild.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ef4/ef4a2c3c0b2f7871f8897cdbf7028407d899f91f/d309829b3c4136c5f4d04b5b822241326605361b.diff) (2723 bytes) - added by therve on 2006-12-14 15:10:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2297 2297
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1166108653000000 1166108653000000
changetime__1177866362000000 1177866362000000
version__None None
owner__therve therve
cc__therve cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun|
|-|-|

Unfortunately I think the tests were a bit incomplete.  I expanded on them somewhat, and testSlots still fails.  testTypeSubclass still passes, however.

Code is in rebuild-newstyle-2297 for now.  Getting one test working is good and worth making the change for, I think.  Maybe we can get the other to work somehow too before merging this, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

I've looked further at this, and I don't manage do find a solution. I found a post of you on python-dev talking about the problem: http://mail.python.org/pipermail/python-list/2003-July/213042.html.

I'm not Python core expert, but I'd say there's definitely a problem in same_slots_added. First the equiv_structs tests is obviously false because of what's is done in compatible_for_assignment. And I didn't understand the size tests, even after reading Martin response to your mail.

I've reduced the problem to this example:
```
class A(object):
    __slots__ = ['a']

a = A()
B = type(A.__name__, A.__bases__, dict(A.__dict__))
a.__class__ = B
```

Maybe python-dev can enlighten us another time ?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I found a solution. But it needs a modification in Python, sorry :). Here's the patch:
```
Index: Objects/typeobject.c
===================================================================
--- Objects/typeobject.c        (revision 53162)
+++ Objects/typeobject.c        (working copy)
@@ -2405,7 +2405,8 @@
 same_slots_added(PyTypeObject *a, PyTypeObject *b)
 {
        PyTypeObject *base = a->tp_base;
-       Py_ssize_t size;
+       Py_ssize_t size, i, size_slot_a, size_slot_b;
+       PyStringObject *elt_a, *elt_b;

        if (base != b->tp_base)
                return 0;
@@ -2416,6 +2417,22 @@
                size += sizeof(PyObject *);
        if (a->tp_weaklistoffset == size && b->tp_weaklistoffset == size)
                size += sizeof(PyObject *);
+       /* Check slots compliance */
+       size_slot_a = PyTuple_GET_SIZE(((PyHeapTypeObject *)a)->ht_slots);
+       size_slot_b = PyTuple_GET_SIZE(((PyHeapTypeObject *)b)->ht_slots);
+       if (size_slot_a != size_slot_b) {
+               return 0;
+       }
+       for (i = 0; i < size_slot_a; i++) {
+               elt_a = (PyStringObject *) PyTuple_GET_ITEM(((PyHeapTypeObject *)a)->ht_slots, i);
+               elt_b = (PyStringObject *) PyTuple_GET_ITEM(((PyHeapTypeObject *)b)->ht_slots, i);
+               if (!(elt_a->ob_size == elt_b->ob_size
+                         && (elt_a->ob_sval[0] == elt_b->ob_sval[0]
+                                 && memcmp(elt_a->ob_sval, elt_b->ob_sval, elt_a->ob_size) == 0))) {
+                       return 0;
+               }
+               size += sizeof(PyObject *);
+       }
        return size == a->tp_basicsize && size == b->tp_basicsize;
 }

```

With this (and another trick in Python, by deleting the descriptors in !__dict!__), I can make the test pass.

I think I'll submit the patch to Python, but we'll have to find a temporary solution.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

My patch went into Python, it should be available in 2.6 (see https://sourceforge.net/tracker/?func=detail&atid=305470&aid=1623563&group_id=5470). I've skipped the test for current version.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @therve|
|-|-|

There should be a sane exception (and tests for it) on Pythons less than 2.6.

The test skip message should also indicate that it is _Python_ 2.6, not Twisted 2.6 (or, say, Linux 2.6 ^_^) where this feature is not yet supported.

Thanks for getting this patch into Python.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I've added a new test, but I'm not really sure on how to set this up. Also, docstrings are awful :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Near the end of the test module, these lines look erroneous:

```
    if sys.hexversion >= 0x02060000:
        test_slots.skip = "__class__ assignment for class with slots should work starting Python 2.6"
```

I guess it's just referring to the wrong test method.

Also, it might be easier to compare sys.version_info to (2, 6), but it's up to you.

In updateInstance, checking the class object for ```__slots__``` would be better than checking the instance object, since an instance might have an ```__slots__}} attribute which won't actually affect anything.  If you feel like fixing that, there should probably be a test case for it, though (and if you don't want to fix it that's fine too).```{__slots__}} attribute which won't actually affect anything.  If you feel like fixing that, there should probably be a test case for it, though (and if you don't want to fix it that's fine too).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

Thanks, I fixed the points.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Cool, looks good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [20116]) Merge rebuild-newstyle-2297

Author: therve
Reviewer: exarkun, glyph
Fixes [#2297](https://github.com/twisted/twisted/issues/2297)

Clean some rough edges of twisted.python.rebuild for new style classes: prevent
the reload of builtin, and gives a sane error for classes with slots. It also
takes into account a recent patch for Python 2.6 that allows rebuild of classes
with slots.
